src/jit/lir.cpp

```diff
@@ -1556,6 +1557,47 @@ bool LIR::Range::CheckLIR(Compiler* compiler, bool checkUnusedValues) const
         }
     }
 
+    // Check lclVar semantics: specifically, ensure that an unaliasable lclVar is not redefined between the
+    // point at which a use appears in linear order and the point at which that use is consumed by its user.
+    // This ensures that it is always safe to treat a lclVar use as happening at the user (rather than at
+    // the lclVar node).
+    //
+    // This happens as a second pass because unused lclVar reads may otherwise appear as outstanding reads
+    // and produce false indications that a write to a lclVar occurs while outstanding reads of that lclVar
+    // exist.
+    SmallHashTable<int, int, 32> unconsumedLclVarReads(compiler);
+    for (GenTree* node : *this)
+    {
+        for (GenTree* operand : node->Operands())
+        {
+            AliasSet::NodeInfo operandInfo(compiler, operand);
+            if (operandInfo.IsLclVarRead())
+            {
+                int count = 0;
+                if (unconsumedLclVarReads.TryRemove(operandInfo.LclNum(), &count) && count > 1)
+                {
+                    unconsumedLclVarReads.AddOrUpdate(operandInfo.LclNum(), count - 1);
+                }
+            }
+        }
+
+        AliasSet::NodeInfo nodeInfo(compiler, node);
+
+        bool unused;
+        if (nodeInfo.IsLclVarRead() && !unusedDefs.TryGetValue(node, &unused))
+        {
+            int count = 0;
+            unconsumedLclVarReads.TryGetValue(nodeInfo.LclNum(), &count);
+            unconsumedLclVarReads.AddOrUpdate(nodeInfo.LclNum(), count + 1);
+        }
+
+        if (nodeInfo.IsLclVarWrite())
+        {
+            int unused;
+            assert(!unconsumedLclVarReads.TryGetValue(nodeInfo.LclNum(), &unused));
```

Comment:
This code could use some comments in general, but especially at the assert, because it's non-obvious. If someone runs into this assert, it would be good to have a comment to help them understand, e.g. "This lclVar has just been written; there had better not be any unconsumed reads of that lclVar from previous nodes."